clang/lib/3C/ConstraintVariables.cpp

```diff
@@ -2115,18 +2115,31 @@ FVComponentVariable::FVComponentVariable(const QualType &QT,
   }
 }
 
-void
-FVComponentVariable::equateWithItype(ProgramInfo &I, bool ForceEquate) const {
+void FVComponentVariable::equateWithItype(
+    ProgramInfo &I, const std::string &ReasonUnchangeable) const {
   Constraints &CS = I.getConstraints();
-  bool IsGeneric = ExternalConstraint->getIsGeneric();
+  const std::string ReasonUnchangeable2 =
+      (ReasonUnchangeable.empty() && ExternalConstraint->getIsGeneric())
+          // TODO: Add a test for this root-cause message somewhere once
+          // diagnostic verification is re-enabled
+          // (https://github.com/correctcomputation/checkedc-clang/issues/503).
+          ? "Internal constraint for generic function declaration, "
+            "for which 3C currently does not support re-solving."
+          : ReasonUnchangeable;
   bool HasBounds = ExternalConstraint->srcHasBounds();
   bool HasItype = ExternalConstraint->srcHasItype();
-  if (HasItype && (ForceEquate || IsGeneric || HasBounds)) {
-    ExternalConstraint->equateWithItype(I, ForceEquate);
+  // If the type cannot change at all (ReasonUnchangeable2 is set), then we
+  // constrain both the external and internal types to not change. Otherwise, if
+  // the variable has bounds, then we don't want the checked (external) portion
+  // of the type to change because that could blow away the bounds, but we still
+  // allow the internal type to change so that the type can change from an itype
+  // to fully checked.
+  if (HasItype && (!ReasonUnchangeable2.empty() || HasBounds)) {
```

Comment:
I think it might be clearer if you introduce a boolean variable to be used here and below e.g.,
```c
bool ItypeCanBeChecked = ReasonUnchangeable2.empty();
```
It's adequately explained in the comment, but having it in the variable name makes it easier for me to read the conditionals.

Reply:
In my view, the following addresses the point more clearly:
```c
bool MustConstrainInternalType = !ReasonUnchangeable2.empty();
```
Is that OK with you?

Reply:
Sure, that sounds fine